tests/python/unittest/test_numpy_op.py

```diff
@@ -418,6 +418,142 @@ def hybrid_forward(self, F, a, b):
+        ((2, 1, 3, 4, 5), (5, 2)),  # Case 5
+        ((1, 3, 5, 4), (1, 4, 3)),  # Case 5
+        ((3, 5, 4), (2, 1, 4, 3)),  # Case 5
+        ((3, 4), (1, 5, 4, 3)),        # Case 5
+    ]
+
+    dtypes = ['float32', 'float64']
+    eps = 1E-4
+    for dtype in dtypes:
+        for hybridize in [True, False]:
+            for shape_a, shape_b in shapes:
+                for grad_req_a in ['write', 'add']:
+                    for grad_req_b in ['write', 'add']:
+                        test_matmul = TestMatmul()
+                        if hybridize:
+                            test_matmul.hybridize()
+                        np_a = _np.random.uniform(-1.0, 1.0, shape_a).astype(dtype)
+                        np_a[abs(np_a) < eps] = 2 * eps
+                        np_b = _np.random.uniform(-1.0, 1.0, shape_b).astype(dtype)
+                        np_b[abs(np_b) < eps] = 2 * eps
+                        a = mx.np.array(np_a, dtype=dtype)
+                        a.attach_grad(grad_req=grad_req_a)
+                        b = mx.np.array(np_b, dtype=dtype)
+                        b.attach_grad(grad_req=grad_req_b)
+
+                        np_out = _np.matmul(np_a, np_b)
+                        with mx.autograd.record():
+                            mx_out = test_matmul(a, b)
+                        assert mx_out.shape == np_out.shape
+                        assert_almost_equal(np_out, mx_out.asnumpy(), rtol=eps, atol=eps)
+                        mx_out.backward()
+                        np_backward = matmul_backward(np_a, np_b)
+                        assert_almost_equal(a.grad.asnumpy(), np_backward[0], rtol = eps, atol=eps)
+                        assert_almost_equal(b.grad.asnumpy(), np_backward[1], rtol = eps, atol=eps)
+                        
+                        # Test imperative once again
+                        mx_out = np.matmul(a, b)
+                        np_out = _np.matmul(np_a, np_b)
+                        assert_almost_equal(mx_out.asnumpy(), np_out, rtol=eps, atol=eps)
+
+    bad_shapes = [
+                ((1,), (2,)),            # mismatched vector vector
+                ((2, 1,), (2,)),         # mismatched matrix vector
+                ((2,), (1, 2)),          # mismatched vector matrix
+                ((1, 2), (3, 1)),        # mismatched matrix matrix
+                ((1,), ()),              # vector scalar
+                ((), (1,)),              # scalar vector
+                ((1, 1), ()),            # matrix scalar
+                ((), (1, 1)),            # scalar matrix
+                ((2, 2, 1), (3, 1, 2)),  # cannot broadcast
+                ]
+
+    for shape_a, shape_b in bad_shapes:
+        a = mx.nd.array(random.random()) if len(shape_a) == 0 else rand_ndarray(shape_a)
+        b = mx.nd.array(random.random()) if len(shape_b) == 0 else rand_ndarray(shape_b)
+        try:
+            mx_res = np.matmul(a.as_np_ndarray(), b.as_np_ndarray())
```

Comment:
you'd better use "assertRaises" for expecting an error being raised.
also with the recommended changes above you can do `np.matmul(a, b)` directly.